x/farming/keeper/staking_test.go

```diff
@@ -134,12 +155,81 @@ func (suite *KeeperTestSuite) TestUnstake() {
 	}
 }
 
+func (suite *KeeperTestSuite) TestUnstakePriority() {
+	// Unstake must withdraw coins from queued staking coins first,
+	// not from already staked coins.
+
+	suite.Stake(suite.addrs[0], sdk.NewCoins(sdk.NewInt64Coin(denom1, 1000000)))
+	suite.AdvanceEpoch()
+
+	check := func(staked, queued int64) {
+		suite.Require().True(coinsEq(sdk.NewCoins(sdk.NewInt64Coin(denom1, staked)), suite.keeper.GetAllStakedCoinsByFarmer(suite.ctx, suite.addrs[0])))
+		suite.Require().True(coinsEq(sdk.NewCoins(sdk.NewInt64Coin(denom1, queued)), suite.keeper.GetAllQueuedCoinsByFarmer(suite.ctx, suite.addrs[0])))
+	}
+
+	suite.Stake(suite.addrs[0], sdk.NewCoins(sdk.NewInt64Coin(denom1, 500000)))
+	check(1000000, 500000)
+
+	suite.Unstake(suite.addrs[0], sdk.NewCoins(sdk.NewInt64Coin(denom1, 300000)))
+	check(1000000, 200000)
+
+	suite.Stake(suite.addrs[0], sdk.NewCoins(sdk.NewInt64Coin(denom1, 500000)))
+	check(1000000, 700000)
+
+	suite.Unstake(suite.addrs[0], sdk.NewCoins(sdk.NewInt64Coin(denom1, 300000)))
+	check(1000000, 400000)
+}
```

Comment:
We can maybe cover a case where we unstake more coins than what we have in `QueueStaking`? 
```go
suite.Unstake(suite.addrs[0], sdk.NewCoins(sdk.NewInt64Coin(denom1, 500000)))
check(900000, 0)
```